### PR TITLE
Allow multiple commands per test in test runner

### PR DESCRIPTION
**Pull request by binji**
This CL allows each test to specify more than one command, each of which
will be run, in order, as long as the previous commands all return
errorcode `0`.

For example:
```
;;; EXE: %(command1)s
;;; FLAGS: --foo=1 --stuff
;;; EXE: %(command2)s
;;; FLAGS: --other-things
```

The goal here is to remove many of the test wrapper scripts (e.g.
`test/run-interp`, `test/run-objdump`, etc.) by having their
functionality represented by multiple commands instead, then wrapped
into a `TOOL` definition.

### CODE REVIEW

File: test/two-commands.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+;;; TOOL: wat2wasm
+;;; FLAGS: -v
```
Comment (sbc100) [inline test/two-commands.txt]: llvm's lit uses "RUN" for this so maybe we should mirror that?  Could/should be separate change I guess.

Comment (sbc100) [reply test/two-commands.txt]: Eventually it might make sense to have flags be part of RUN too, no need for a separate line is there?

File: test/dump/invalid-data-segment-no-memory.txt [hunk 2]
```diff
@@ -1,3 +1,4 @@
+;;; TOOL: wat2wasm
```
Comment (sbc100) [inline test/dump/invalid-data-segment-no-memory.txt]: So part of this change to remove the default tool?  Perhaps that could be a separate CL since it seems somewhat orthogonal to supporting multiple tools.

File: test/run-tests.py [hunk 3]
```diff
@@ -265,11 +327,16 @@ def CreateRoundtripInfo(self, fold_exprs):
     result.input_ = self.input_
     result.expected_stdout = ''
     result.expected_stderr = ''
+
     # TODO(binji): It's kind of cheesy to keep the enable flag based on prefix.
     # Maybe it would be nicer to add a new directive ENABLE instead.
-    result.flags = [flag for flag in self.flags if flag.startswith('--enable')]
+    old_cmd = self.cmds[0]
+    new_cmd = result.cmds[0]
+    new_cmd.flags = [flag for flag in old_cmd.flags
```
Comment (sbc100) [inline test/run-tests.py]: i think its ok to use single char variable names in list comprehensions, especially if it means not having to wrap them.

File: test/run-tests.py [hunk 4]
```diff
@@ -413,32 +478,9 @@ def Parse(self, filename):
 
     # If the test didn't specify a executable (either via EXE or indirectly
     # via TOOL, then use the default tool)
-    if not self.exe:
+    if not self.cmds:
```
Comment (sbc100) [inline test/run-tests.py]: So there is still a default?

File: test/run-tests.py [hunk 5]
```diff
@@ -189,7 +199,8 @@ def Get(self):
     return self.value[0]
 
 
-def RunCommandWithTimeout(command, cwd, timeout, console_out=False, env=None):
+def RunCommandListWithTimeout(command, cwd, timeout, console_out=False,
```
Comment (sbc100) [inline test/run-tests.py]: Why rename this?  It still takes a single command doesn't it?

File: test/run-tests.py [hunk 6]
```diff
@@ -237,6 +248,52 @@ def KillProcess(timeout=True):
   return stdout, stderr, returncode, duration
 
 
+def AsFlagsList(value):
```
Comment (sbc100) [inline test/run-tests.py]: I prefer `args` to `flags` in general.  Maybe "SplitArgs"?

File: test/run-tests.py [hunk 7]
```diff
@@ -597,15 +641,30 @@ def RunTest(info, options, variables, verbose_level=0):
...
+  total_duration = 0
 
-  try:
-    return RunCommandWithTimeout(cmd, cwd, timeout, verbose_level > 0, env)
-  except (Error, KeyboardInterrupt) as e:
-    return e
+  for cmd in info.cmds:
+    cmd_list = cmd.AsList(variables, options.arg, verbose_level)
+    if options.print_cmd:
+      print(' '.join(cmd_list))
```
Comment (sbc100) [inline test/run-tests.py]: How about cmd.Print()  or `print cmd` then implement Command.__str__()?

File: test/run-tests.py [hunk 8]
```diff
@@ -597,15 +641,30 @@ def RunTest(info, options, variables, verbose_level=0):
...
+  total_stderr = ''
+  returncode = 0
+  total_duration = 0
 
-  try:
-    return RunCommandWithTimeout(cmd, cwd, timeout, verbose_level > 0, env)
-  except (Error, KeyboardInterrupt) as e:
-    return e
+  for cmd in info.cmds:
+    cmd_list = cmd.AsList(variables, options.arg, verbose_level)
```
Comment (sbc100) [inline test/run-tests.py]: I don't really like the name `AsList` but I can't think of a better one right now.  Something like ".Get()" or ".Format()"?  I don't really like those either :)

File: test/run-tests.py [hunk 9]
```diff
@@ -597,15 +641,30 @@ def RunTest(info, options, variables, verbose_level=0):
...
+      stdout, stderr, returncode, duration = (
+          RunCommandListWithTimeout(cmd_list, cwd, timeout,
+                                    verbose_level > 0, env))
+      total_stdout += stdout
+      total_stderr += stderr
+      total_duration += duration
+
+      if returncode != 0:
+        break
+    except (Error, KeyboardInterrupt) as e:
```
Comment (sbc100) [inline test/run-tests.py]: Move the accept up so it only surrounds the RunCommandListWithTimeout call?

File: test/dump/invalid-data-segment-no-memory.txt (hunk 2 above)
Comment (binji) [reply test/dump/invalid-data-segment-no-memory.txt]: No, it still has a default tool, but it doesn't work with just `FLAGS`. I guess it would be better to just set it as the default if the `FLAGS` directive is found.

File: test/run-tests.py (hunk 5 above)
Comment (binji) [reply test/run-tests.py]: Yeah, good point. I wanted to differentiate between passing a `Command` object and a list of strings (`CommandList`) but maybe it's not worth it.

File: test/run-tests.py (hunk 6 above)
Comment (binji) [reply test/run-tests.py]: Me too, but I wanted to do the flags -> args rename in another CL

File: test/run-tests.py (hunk 3 above)
Comment (binji) [reply test/run-tests.py]: Done.

File: test/run-tests.py (hunk 9 above)
Comment (binji) [reply test/run-tests.py]: Done.

File: test/two-commands.txt (hunk 1 above)
Comment (binji) [reply test/two-commands.txt]: Yeah, I want to do that too (move FLAGS into RUN)

**Comment by binji** [comment]
OK, this should be a bit cleaner, PTAL!
